cmd/outline-ss-server/main.go

```diff
@@ -122,65 +111,81 @@ func (s *SSServer) loadConfig(filename string) error {
 		return fmt.Errorf("failed to load config (%v): %w", filename, err)
 	}
 
-	portChanges := make(map[int]int)
-	portCiphers := make(map[int]*list.List) // Values are *List of *CipherEntry.
-	for _, keyConfig := range config.Keys {
-		portChanges[keyConfig.Port] = 1
-		cipherList, ok := portCiphers[keyConfig.Port]
+	// We hot swap the config by having the old and new listeners both live at
+	// the same time. This means we create listeners for the new config first,
+	// and then close the old ones after.
+	oldListenerSet := s.lnSet
+	s.lnSet = s.lnManager.NewListenerSet()
+	var totalCipherCount int
+
+	portHandlers := make(map[int]Handler)
+	for _, legacyKeyServiceConfig := range config.Keys {
+		handler, ok := portHandlers[legacyKeyServiceConfig.Port]
 		if !ok {
-			cipherList = list.New()
-			portCiphers[keyConfig.Port] = cipherList
+			handler = &connHandler{
+				ciphers:     service.NewCipherList(),
+				tcpTimeout:  tcpReadTimeout,
+				natTimeout:  s.natTimeout,
+				m:           s.m,
+				replayCache: &s.replayCache,
+			}
+			portHandlers[legacyKeyServiceConfig.Port] = handler
 		}
-		cryptoKey, err := shadowsocks.NewEncryptionKey(keyConfig.Cipher, keyConfig.Secret)
+		cryptoKey, err := shadowsocks.NewEncryptionKey(legacyKeyServiceConfig.Cipher, legacyKeyServiceConfig.Secret)
 		if err != nil {
-			return fmt.Errorf("failed to create encyption key for key %v: %w", keyConfig.ID, err)
+			return fmt.Errorf("failed to create encyption key for key %v: %w", legacyKeyServiceConfig.ID, err)
 		}
-		entry := service.MakeCipherEntry(keyConfig.ID, cryptoKey, keyConfig.Secret)
-		cipherList.PushBack(&entry)
-	}
-	for port := range s.ports {
-		portChanges[port] = portChanges[port] - 1
+		entry := service.MakeCipherEntry(legacyKeyServiceConfig.ID, cryptoKey, legacyKeyServiceConfig.Secret)
+		handler.AddCipher(&entry)
 	}
-	for portNum, count := range portChanges {
-		if count == -1 {
-			if err := s.removePort(portNum); err != nil {
-				return fmt.Errorf("failed to remove port %v: %w", portNum, err)
-			}
-		} else if count == +1 {
-			if err := s.startPort(portNum); err != nil {
-				return err
+	for portNum, handler := range portHandlers {
+		totalCipherCount += handler.NumCiphers()
+		for _, network := range []string{"tcp", "udp"} {
```

Comment:
Rather than being generic, let's stay strongly typed if you can.

I imagine the loop being:

- for each service
  - create Stream (if there are stream listeners) and Packet handlers (if there are packet listeners)
  - for each listener
     - start the listener, and call StreamServe or PacketServe on the listener and handler, based on the listener type.

Perhaps create the Stream and Packet handlers on demand when you first need them, as you go over the listeners.